### PR TITLE
Allow installing plugins from path via CLI

### LINKED ISSUE #6953: `bundle plugin uninstall` wipes out your code if using a path
### Describe the problem as clearly as you can

In my Gemfile, I have something like

`plugin "my_plugin", path: "../my_plugin"`

I was testing some installation behavior of my plugin, and so needed to re-install it. First I ran `bundle plugin uninstall my_plugin` to clean it out of the plugin index. Then to my horror, I realized that the entire my_plugin directory was gone!

### Did you try upgrading rubygems & bundler?

Yes. I'm on rubygems 3.4.19, and bundler 2.4.19.

### What were you expecting to happen?

Because it's a `path:` reference in the gemfile, I expected it to only get removed from .bundle/plugins/index, but the actual plugin (which is not cached in .bundle) would remain untouched.

### What actually happened?

My local code was removed!

## Environment

```
Bundler       2.4.19
  Platforms   ruby, arm64-darwin-21
Ruby          3.1.3p185 (2022-11-24 revision 1a6b16756e0ba6b95ab71a441357ed5484e33498) [arm64-darwin-21]
  Full Path   /Users/cody/.rubies/ruby-3.1.3/bin/ruby
  Config Dir  /Users/cody/.rubies/ruby-3.1.3/etc
RubyGems      3.4.19
  Gem Home    /Users/cody/.gem/ruby/3.1.3
  Gem Path    /Users/cody/.gem/ruby/3.1.3:/Users/cody/.rubies/ruby-3.1.3/lib/ruby/gems/3.1.0
  User Home   /Users/cody
  User Path   /Users/cody/.gem/ruby/3.1.0
  Bin Dir     /Users/cody/.gem/ruby/3.1.3/bin
Tools         
  Git         2.39.2 (Apple Git-143)
  RVM         not installed
  rbenv       not installed
  chruby      0.3.9
```

## Bundler Build Metadata

```
Built At          2023-08-17
Git SHA           86f98098e3
Released Version  true
```

## Bundler settings

```
gem.changelog
  Set for the current user (/Users/cody/.bundle/config): false
gem.ci
  Set for the current user (/Users/cody/.bundle/config): false
gem.coc
  Set for the current user (/Users/cody/.bundle/config): false
gem.linter
  Set for the current user (/Users/cody/.bundle/config): "rubocop"
gem.mit
  Set for the current user (/Users/cody/.bundle/config): false
gem.test
  Set for the current user (/Users/cody/.bundle/config): false
gemfile
  Set via BUNDLE_GEMFILE: "/Users/cody/src/canvas-lms/Gemfile"
unset
  Set for the current user (/Users/cody/.bundle/config): "frozen"
```

## Gemfile

```
<snip>
```

### PR DESCRIPTION
**Pull request by ccutrer**
Refs #3319

## What was the end-user or developer problem that led to this PR?

You can install a local plugin gem in the Gemfile via `plugin 'plugin', path: '../plugin'`, but you cannot do the analogous `bundle plugin install plugin --path=../plugin` from the CLI.

## What is your fix for the problem, implemented in this PR?

Refactor how sources are passed to the plugin installer, support the path source, remove duplicate code for git sources, and update docs.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by ccutrer** [comment]
I rebased, and removed the accidental change to the rubocop tools lockfile.

**Comment by deivid-rodriguez** [comment]
You just need a couple more tweaks to get CI green. Run `rake update_manifest`, `rake man:build`, and amend the latest commit with the changes generated.

Also, I would probably extract the bug fix to a separate PR and leave this to the `--path` CLI flag addition?

**Comment by ccutrer** [comment]
Done. Bugfix split off to #7001 (For some reason I was thinking it depended on this commit, so put both commits in the same PR).

**Comment by ccutrer** [comment]
@deivid-rodriguez: any further comments on this PR?

File: bundler/lib/bundler/plugin/installer.rb
Comment (deivid-rodriguez) [inline bundler/lib/bundler/plugin/installer.rb:49]: It the `--local_git` option now a loop? I guess we should keep it working as before for now.

**Review by deivid-rodriguez — COMMENTED** [review]
Looks really good, just have some comments on the `--local_git` option.

File: bundler/lib/bundler/plugin/installer.rb
Comment (ccutrer) [reply bundler/lib/bundler/plugin/installer.rb:49]: it simply renames the option to :git here, before choosing the install method. the old install_git and install_local_git methods were identical (except choosing a different option key). so actual functionality has not changed

Comment (deivid-rodriguez) [reply bundler/lib/bundler/plugin/installer.rb:49]: Oh, my bad, I read the code incorrectly, thanks for clarifying!

File: bundler/spec/plugins/install_spec.rb
Comment (deivid-rodriguez) [inline bundler/spec/plugins/install_spec.rb:96]: This is another different bug fix / change included with this PR, right? Previously we would interpret `--source` as a `git` source if `branch` or `ref` were specified. Now we're more strict about it.

I'm in general in favor of this since it makes plugin sources more consistent with Gemfile sources, but since this was coded as specs, maybe extract a PR & discuss/review there?

**Review by deivid-rodriguez — COMMENTED** [review]
The PR looks good in general, but there's a few different things going on, and while I like all of them, I think perhaps the change about `--source` in combination with `--branch`, or `--ref` could be extracted as a separate PR. Particularly since it was codified in specs so maybe someone wanted this for a reason.

**Comment by deivid-rodriguez** [comment]
@ccutrer Apologies for that extra nit where I had basically approved this earlier. Just noticed this in the last review pass when I was going to merge it. I'm also happy to do this extraction myself if you prefer!

**Comment by ccutrer** [comment]
Not a problem. I'll hopefully get around to it on Monday or Tuesday. I've been super busy updating lots of gems since the Rails 7.1 release. I wanted to to re-look at this to understand if it actually made sense to do `--source ... --branch ...`.

File: bundler/spec/plugins/install_spec.rb
Comment (ccutrer) [reply bundler/spec/plugins/install_spec.rb:96]: I've actually broken this into 4 commits now, but they all depend on each other. I don't know if you have a preference to just have 4 commits in this PR, or one PR per commit. #7045 is the first commit in the stack if you prefer separate PRs; I'll create the next PR once that's merged, etc.

**Comment by ccutrer** [comment]
I actually split this into 4 atomic commits (that are all dependent on each other to avoid conflicts). I'm not sure if you have a preference for 1-commit-per-PR or not. If so, #7045 is the first commit in the stack, and as each one is submitted, I'll open the next PR. For now the branch this PR is from has all the commits, and I'll rebase it once its predecessors have been submitted separately.

**Comment by deivid-rodriguez** [comment]
Wow, that's so nice of you. Generally I don't mind small extra changes sneaked into PR, but since you split this already, separate PRs are indeed better because they will show up as different improvements in our release notes. Thanks for doing this, I'll start with the docs PR!

**Comment by ccutrer** [comment]
#7048 is the second in the stack

**Comment by ccutrer** [comment]
Rebased and re-worked spec to use gemfile installation, so this PR no longer has any dependencies.

**Comment by deivid-rodriguez** [comment]
I merged the deprecation of `bundle plugin install --local_git`. https://github.com/rubygems/rubygems/pull/6960/commits/f44322cad11e5d2507b367a35f7eb78f6e564440 can be extracted now to a separate PR.

**Comment by deivid-rodriguez** [comment]
I extracted last PR into #7529.

**Comment by deivid-rodriguez** [comment]
I merged #7529 and gave this PR one last rebase!

**Review by deivid-rodriguez — APPROVED** [review]
Alright, let's wrap this up. Thanks so much for all the improvements @ccutrer!

**Comment by ccutrer** [comment]
This has caused a regression in Bundler 2.5.8 that plugin ..., path: ... in a Gemfile can no longer find the plugin. Investigating now...

**Comment by deivid-rodriguez** [comment]
Thanks for the super fast report and initial investigation. I can get a quick release out once we fix the problem.

**Pull request merged**
